### PR TITLE
Provides IO trace events handler to work with parsed IO

### PR DESCRIPTION
**Pull request by mariuszbarczak**
The parsed IO contains basic IO information (LBA, length, etc).
It is supplemented by related information like filesystem one.
In addition it provides post parse information (latency,
queue depth, etc...)

Signed-off-by: Mariusz Barczak <mariusz.barczak@intel.com>

### CODE REVIEW

File: source/octf/trace/parser/IoTraceEventHandler.cpp [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        // This event handler presents traces from time '0', we remember
+        // first event timestamp and subtract by this value for each event
+        m_timestampOffset = timestamp;
+    }
+    traceEvent->mutable_header()->set_timestamp(timestamp - m_timestampOffset);
+
+    switch (traceEvent->EventType_case()) {
+    case Event::EventTypeCase::kDeviceDescription: {
+        // Remember device
+        auto device = traceEvent->mutable_devicedescription();
```
Comment (tomaszrybicki) [inline source/octf/trace/parser/IoTraceEventHandler.cpp]: I don't think we need mutable getter, I'd use a normal one for clarity.

File: source/octf/trace/parser/IoTraceEventHandler.h [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+    virtual std::shared_ptr<proto::trace::Event> getEventMessagePrototype()
+            override;
+
+    void flushEvetns();
+
+private:
+    typedef proto::trace::ParsedEvent ParsedEvent;
+    std::map<uint64_t, ParsedEvent> m_cache;
+    std::map<uint64_t, proto::trace::EventDeviceDescription> m_devices;
```
Comment (tomaszrybicki) [inline source/octf/trace/parser/IoTraceEventHandler.h]: I suggest adding const to DeviceDescription

File: source/octf/trace/parser/IoTraceEventHandler.cpp [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        m_devices[device->id()] = *device;
+    } break;
+
+    case Event::EventTypeCase::kIo: {
+        auto sid = traceEvent->mutable_header()->sid();
```
Comment (tomaszrybicki) [inline source/octf/trace/parser/IoTraceEventHandler.cpp]: Please use non mutable getter

File: source/octf/trace/parser/IoTraceEventHandler.cpp [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+            if (iter->first > sid) {
+                // Iterating over higher SID, what means over never event,
+                // so
+                break;
+            }
+
+            if (iter->second.mutable_io()->lba() == lba &&
+                iter->second.mutable_io()->len() == len) {
+                uint64_t latency =
+                        timestamp - iter->second.mutable_header()->timestamp();
```
Comment (tomaszrybicki) [inline source/octf/trace/parser/IoTraceEventHandler.cpp]: Please use normal getter

File: source/octf/trace/parser/IoTraceEventHandler.cpp [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        flushEvetns();
+    } break;
+
+    case Event::kFilesystemMeta: {
+        auto sid = traceEvent->mutable_filesystemmeta()->refsid();
+
+        auto iter = m_cache.find(sid);
+        if (iter != m_cache.end()) {
+            auto &dst = *iter->second.mutable_file();
+            const auto &src = *traceEvent->mutable_filesystemmeta();
```
Comment (tomaszrybicki) [inline source/octf/trace/parser/IoTraceEventHandler.cpp]: Please use normal getter

File: source/octf/trace/parser/IoTraceEventHandler.cpp [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+                iter->second.mutable_io()->set_error(error);
+
+                break;
+            }
+        }
+        flushEvetns();
+    } break;
+
+    case Event::kFilesystemMeta: {
+        auto sid = traceEvent->mutable_filesystemmeta()->refsid();
```
Comment (tomaszrybicki) [inline source/octf/trace/parser/IoTraceEventHandler.cpp]: Please use normal getter

File: source/octf/trace/parser/IoTraceEventHandler.cpp [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+std::shared_ptr<proto::trace::Event>
+IoTraceEventHandler::getEventMessagePrototype() {
+    return std::make_shared<proto::trace::Event>();
+}
+
+void IoTraceEventHandler::flushEvetns() {
+    auto iter = m_cache.begin();
+    auto end = m_cache.end();
+    for (; iter != end;) {
+        if (iter->second.mutable_io()->latency()) {
```
Comment (tomaszrybicki) [inline source/octf/trace/parser/IoTraceEventHandler.cpp]: Please use normal getter.
Please also provide comment of what is this 'if' checking (checking for completion presence?)

File: source/octf/trace/parser/IoTraceEventHandler.cpp [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+            handleIO(iter->second);
```
Comment (tomaszrybicki) [inline source/octf/trace/parser/IoTraceEventHandler.cpp]: Please add information in class documentation that IOs are handled in order of their requests, not completions.

File: source/octf/trace/parser/IoTraceEventHandler.cpp [hunk 9]
```diff
@@ -0,0 +1,142 @@
+/*
```
Comment (tomaszrybicki) [inline source/octf/trace/parser/IoTraceEventHandler.cpp]: Overall very clean and simple implementation :+1:

File: source/octf/trace/parser/IoTraceEventHandler.h (hunk 2 above)
Comment (mariuszbarczak) [reply source/octf/trace/parser/IoTraceEventHandler.h]: I think it's not needed

**Pull request merged**